### ISSUE TITLE
Batch normalization layer slows convergence

### ISSUE THREAD

**Issue by daiwei89** [tflearn/tflearn#530]
In all my experiences using my own implementation, batch normalization help convergence. However, that's not the case for tflearn's batch normalization. Use the convnet_mnist.py (https://github.com/tflearn/tflearn/blob/master/examples/images/convnet_mnist.py) as an example. When I insert batch_normalization layers (see gist code https://gist.github.com/daiwei89/a0d9600050003249e7c30f8e63742985) then the training loss (disregarding the val_loss) decreases a lot slower. This is the same in my other experiments, where I have carefully maintained `tflearn.is_training(True/False)` during training/validation stages. Am I missing something?

Here's the loss output of convnet_mnist.py with and without the inserted batch normalization layers.

## Without batch normalization layers:
Training Step: 100  | total loss: 0.30543 | time: 2.792s
| Adam | epoch: 001 | loss: 0.30543 - acc: 0.9117 | val_loss: 0.19105 - val_acc: 0.9415 -- iter: 06400/55000
--
Training Step: 200  | total loss: 0.18258 | time: 5.079s
| Adam | epoch: 001 | loss: 0.18258 - acc: 0.9393 | val_loss: 0.11066 - val_acc: 0.9678 -- iter: 12800/55000
--
Training Step: 300  | total loss: 0.17067 | time: 7.356s
| Adam | epoch: 001 | loss: 0.17067 - acc: 0.9523 | val_loss: 0.14398 - val_acc: 0.9590 -- iter: 19200/55000
--
Training Step: 400  | total loss: 0.16256 | time: 9.656s
| Adam | epoch: 001 | loss: 0.16256 - acc: 0.9533 | val_loss: 0.09170 - val_acc: 0.9752 -- iter: 25600/55000
--
Training Step: 500  | total loss: 0.15717 | time: 11.946s
| Adam | epoch: 001 | loss: 0.15717 - acc: 0.9595 | val_loss: 0.13721 - val_acc: 0.9590 -- iter: 32000/55000
--
Training Step: 600  | total loss: 0.16846 | time: 14.235s
| Adam | epoch: 001 | loss: 0.16846 - acc: 0.9512 | val_loss: 0.15143 - val_acc: 0.9581 -- iter: 38400/55000
--
Training Step: 700  | total loss: 0.18762 | time: 16.502s
| Adam | epoch: 001 | loss: 0.18762 - acc: 0.9514 | val_loss: 0.08024 - val_acc: 0.9768 -- iter: 44800/55000
--
Training Step: 800  | total loss: 0.23835 | time: 18.786s
| Adam | epoch: 001 | loss: 0.23835 - acc: 0.9433 | val_loss: 0.13808 - val_acc: 0.9642 -- iter: 51200/55000
--
Training Step: 860  | total loss: 0.24259 | time: 20.669s
| Adam | epoch: 001 | loss: 0.24259 - acc: 0.9388 | val_loss: 0.13082 - val_acc: 0.9670 -- iter: 55000/55000
--
Training Step: 900  | total loss: 0.10989 | time: 1.647s
| Adam | epoch: 002 | loss: 0.10989 - acc: 0.9773 | val_loss: 0.09355 - val_acc: 0.9754 -- iter: 02560/55000
--
Training Step: 1000  | total loss: 0.13950 | time: 3.908s
| Adam | epoch: 002 | loss: 0.13950 - acc: 0.9619 | val_loss: 0.09803 - val_acc: 0.9708 -- iter: 08960/55000

## With batch normalization layers:
Training Step: 100  | total loss: 0.67903 | time: 3.710s
| Adam | epoch: 001 | loss: 0.67903 - acc: 0.7875 | val_loss: 0.93544 - val_acc: 0.6710 -- iter: 06400/55000
--
Training Step: 200  | total loss: 0.47235 | time: 6.551s
| Adam | epoch: 001 | loss: 0.47235 - acc: 0.8528 | val_loss: 0.32006 - val_acc: 0.9032 -- iter: 12800/55000
--
Training Step: 300  | total loss: 0.38594 | time: 9.389s
| Adam | epoch: 001 | loss: 0.38594 - acc: 0.8805 | val_loss: 0.28783 - val_acc: 0.9146 -- iter: 19200/55000
--
Training Step: 400  | total loss: 0.33900 | time: 12.201s
| Adam | epoch: 001 | loss: 0.33900 - acc: 0.8959 | val_loss: 0.28479 - val_acc: 0.9092 -- iter: 25600/55000
--
Training Step: 500  | total loss: 0.38939 | time: 15.023s
| Adam | epoch: 001 | loss: 0.38939 - acc: 0.8848 | val_loss: 0.24028 - val_acc: 0.9227 -- iter: 32000/55000
--
Training Step: 600  | total loss: 0.24920 | time: 17.844s
| Adam | epoch: 001 | loss: 0.24920 - acc: 0.9239 | val_loss: 0.19104 - val_acc: 0.9425 -- iter: 38400/55000
--
Training Step: 700  | total loss: 0.30217 | time: 20.664s
| Adam | epoch: 001 | loss: 0.30217 - acc: 0.9054 | val_loss: 0.28573 - val_acc: 0.9088 -- iter: 44800/55000
--
Training Step: 800  | total loss: 0.35916 | time: 23.512s
| Adam | epoch: 001 | loss: 0.35916 - acc: 0.8898 | val_loss: 0.30159 - val_acc: 0.9009 -- iter: 51200/55000
--
Training Step: 860  | total loss: 0.35106 | time: 25.718s
| Adam | epoch: 001 | loss: 0.35106 - acc: 0.8929 | val_loss: 0.25578 - val_acc: 0.9176 -- iter: 55000/55000
--
Training Step: 900  | total loss: 0.30839 | time: 1.908s
| Adam | epoch: 002 | loss: 0.30839 - acc: 0.9046 | val_loss: 0.26711 - val_acc: 0.9187 -- iter: 02560/55000
--
Training Step: 1000  | total loss: 0.26333 | time: 4.744s
| Adam | epoch: 002 | loss: 0.26333 - acc: 0.9164 | val_loss: 0.20096 - val_acc: 0.9389 -- iter: 08960/55000

**Comment by hiranumn** [comment]
Is there any follow up on this issue? I also seem to have the same problem when I experiment with AlexNet w/ batch norm and AlexNet w/o batch norm.

**Comment by cvnerds** [comment]
Did you try increasing the learning rate? When I read the paper I thought the point being was that batch normalisation can make use of learning rates that are larger by an order of magnitude (7x-50x) without sacrificing accuracy. See also my comments on the gist. However, I'm no expert.

**Comment by wssrcok** [comment]
same issue on my MNIST cnn built from scratch
